samples: posix: implement philosphers using pthreads

cfriedt opened the pull request:
This sample takes a POSIX-y spin on the existing Dining Philosophers sample application, similarly to how we have extended it for [CMSIS v1](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/portability/cmsis_rtos_v1/philosophers) and [CMSIS v2](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/portability/cmsis_rtos_v2).

The objects used in the POSIX version are of type `pthread_mutex_t`, and the threads are of type `pthread_t`.

Commit:
posix: internal: make priority / policy transforms available

Two functions can be used with relative ease to convert between
Zephyr and POSIX priorities and policies. Namely,

uint32_t zephyr_to_posix_priority(int32_t z_prio, int *policy)
int32_t posix_to_zephyr_priority(uint32_t priority, int policy)

These are not necessarily public API, but they helped with the
POSIX Philosophers Sample, which is in a subsequent commit.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
samples: posix: implement philosphers using pthreads

This sample takes a POSIX-y spin on the existing
Dining Philosophers sample application. The objects
used in the POSIX version are pthread_mutex_t, and
the threads are pthread_t.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
Really can't win with these compliance failure false positives 😅 
```cpp
BUILD_ASSERT(CONFIG_MAX_PTHREAD_COUNT == CONFIG_MAX_PTHREAD_MUTEX_COUNT - 1);
```
```shell
CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
File:samples/posix/philosophers/src/main.c
Line:22
```

ycsin commented on samples/posix/philosophers/src/main.c:
```diff
@@ -0,0 +1,262 @@
+	/* Wait a few seconds before main() exit, giving the sample the
+	 * opportunity to dump some output before coverage data gets emitted
+	 */
+	k_sleep(K_MSEC(5000));
```
```suggestion
	sleep(5);
```

cfriedt replied:
Fixed!

ycsin commented on samples/posix/philosophers/src/main.c:
```diff
@@ -0,0 +1,262 @@
+
+		delay = get_random_delay(my_id, 25);
+		print_phil_state(my_id, "  EATING  [ %s%d ms ] ", delay);
+		k_msleep(delay);
```
```suggestion
		usleep(delay * USEC_PER_MSEC);
```

cfriedt replied:
Fixed!

ycsin reviewed (commented):
Is it necessary for the POSIX version to closely resemble the standard Zephyr one? Otherwise we do not need to publicize the internal functions and make this sample more 'POSIX'

cfriedt commented on samples/posix/philosophers/src/main.c:
```diff
@@ -0,0 +1,262 @@
+	 * and the current uptime to create some pseudo-randomness. It produces
+	 * a value between 0 and 31.
+	 */
+	int32_t delay = (k_uptime_get_32() / 100 * (id + 1)) & 0x1f;
```
Meh, this is maybe overkill. I think uptime is mainly just some arbitrary number. The POSIX version might use something like `rand_r()`, but it's not a big issue.

cfriedt replied:
I guess it makes sense to use POSIX functions mostly in this sample.

cfriedt replied:
Fixed!

cfriedt commented on samples/posix/philosophers/src/main.c:
```diff
@@ -0,0 +1,262 @@
+#elif defined(CONFIG_PREEMPT_ENABLED)
+	return phil;
+#else
+#error unpossible
```
it's unpossible in the original. I'm tempted to leave it there for posterity 😁

cfriedt commented on samples/posix/philosophers/src/main.c:
```diff
@@ -0,0 +1,262 @@
+static int new_prio(int phil)
+{
+#if defined(CONFIG_COOP_ENABLED) && defined(CONFIG_PREEMPT_ENABLED)
+#if SAME_PRIO
```
probably not

cfriedt replied:
Ok - I added a flag for it in the `testcase.yaml`

Commit:
scripts: checkpatch.pl: fix constant comparison false positives

Comparisons that have constants on both side of the operator
were getting flagged incorrectly. Adjust the check so that
pure constant comparisons are not flagged, reducing false
positives.

WARNING:CONSTANT_COMPARISON: \
  Comparisons should place the constant on the right side of \
  the test
+BUILD_ASSERT(CONFIG_MAX_PTHREAD_COUNT == \
  CONFIG_MAX_PTHREAD_MUTEX_COUNT - 1);

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

kartben commented on samples/posix/philosophers/README.rst:
```diff
@@ -0,0 +1,57 @@
+Building and Running
+********************
+
+This project outputs to the console.  It can be built and executed
```
Line wrapping looks very aggressive here (and inconsistent throughout the file)

cfriedt replied:
Fixed!

kartben commented on samples/posix/philosophers/prj.conf:
```diff
@@ -0,0 +1,25 @@
+#Enable thread awareness for debugging tools supporting it
+CONFIG_DEBUG_THREAD_INFO=y
+
+# CONFIG_SAMPLE_ERROR_CHECKING=y
+# CONFIG_PTHREAD_MUTEX_LOG_LEVEL_DBG=y
+# CONFIG_PTHREAD_LOG_LEVEL_DBG=y
```
Leftover?

cfriedt replied:
Well, the original sample had a commented config option for debug purposes. I thought I would leave these here as well as hints in case someone needs them.

Actually, no debugging was necessary for this - it Just Worked (TM)

kartben reviewed (changes requested):
Would be great to have a chance to review if you are going to make me the assignee :)

cfriedt commented:
> Would be great to have a chance to review if you are going to make me the assignee :)

I was auto assigned. You were added as a reviewer for weeks. It didn't seem like you were interested to be honest.

nashif commented:
> I was auto assigned. You were added as a reviewer for weeks. It didn't seem like you were interested to be honest.

why did you re-assign to @kartben just before merging though?

cfriedt commented on samples/posix/philosophers/sample.yaml:
```diff
@@ -0,0 +1,21 @@
+  name: POSIX Philosophers
+common:
+  tags:
+    - inroduction
```
I'll fix this in the original as well (that was where the typo came from).

cfriedt commented on samples/posix/philosophers/Kconfig:
```diff
@@ -0,0 +1,16 @@
+	default y
+
+config SAMPLE_SAME_PRIO
+	bool "Print debug information"
```
Fixed

cfriedt commented:
> > I was auto assigned. You were added as a reviewer for weeks. It didn't seem like you were interested to be honest.
> 
> why did you re-assign to @kartben just before merging though?

Mainly to elicit a review. There were a lot of reviewers assigned over the last two weeks and I only received 1, in spite of asking 3 times in `#pr-help` over the last two weeks.

cfriedt commented:
@kartben - fixes in https://github.com/zephyrproject-rtos/zephyr/pull/68188 for you to review.
